A6.md

```diff
@@ -0,0 +1,465 @@
+
+![Where Retries Occur](A6_graphics/WhereRetriesOccur.png)
+
+[Link to SVG file](A6_graphics/WhereRetriesOccur.svg)
+
+#### When Retries are Valid
+
+In certain cases it is not valid to retry an RPC. These cases occur when the RPC has been *committed*, and thus it does not make sense to perform the retry.
+
+An RPC becomes *committed* in two scenarios:
+
+1. The client receives a non-error response (either an explicit OK status or any response message) from the server.
+2. The client’s outgoing message has overflowed the gRPC client library’s buffer.
+
+The reasoning behind the first scenario is that as soon as a non-error response is seen from the server, it is transmitted up to the client application. This may fundamentally change the state of the client, so we cannot safely retry if a failure occurs later in the RPC’s life. This design feature of retries necessitates a small change to the gRPC wire protocol, which is described in the [following section](#change-to-grpc-wire-format)
+
+To clarify the second scenario, we define an *outgoing message* as everything the client sends on its connection to the server. For unary and server streaming calls, the outgoing message is a single message. For client and bidirectional streaming calls, the outgoing message is the entire message stream issued by the client after opening the connection. The gRPC client library buffers outgoing messages, and as long as the entirety of the outgoing message is in the buffer, it can be resent and retried. But as soon as the outgoing message grows too large to buffer, the gRPC client library cannot replay the entire stream of messages, and thus retries are not valid.
+
+#### Change to gRPC Wire Format
+
+If a server's response only consists of a status code, it is currently legal to send the status back in the trailers. Based on the previous section, this would mean that those RPCs would not be considered valid to retry, since the client would already have seen the initial metadata.
+
+While implementing retries we will make the wire protocol slightly stricter, and ensure that the status is sent back in the initial metadata if that is all the server is going to send, which will give the RPC a chance of being retried if it should.
+
+#### Memory Management (Buffering)
+
+The gRPC client library will support application-configured limits for the amount of memory used for retries. It is suggested that the client sets `retry_buffer_size_in_bytes` to limit the total amount of memory used to buffer retryable or hedged RPCs. The client should also set `per_rpc_buffer_limit_in_bytes` to limit the amount of memory used by any one RPC (to prevent a single large RPC from using the whole buffer and preventing retries of subsequent smaller RPCs). These limits are configured by the client, rather than coming from the service config.
```

Comment:
I believe that this could be made significantly simpler by saying: number of buffered per RPC.
This makes it easier for users of the library to reason about it: they know that 5 unary RPCs will make becasue retries buffer 10.